crates/wast/src/ast/assert_expr.rs

```diff
@@ -0,0 +1,161 @@
+use crate::ast::{kw, Float32, Float64};
+use crate::parser::{Parse, Parser, Result};
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub enum AssertExpression {
+    I32(i32),
+    I64(i64),
+    F32(NanPattern<Float32>),
+    F64(NanPattern<Float64>),
+    V128(V128Pattern),
+}
+
+impl <'a> Parse<'a> for AssertExpression {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let keyword = parser.step(|c| {
+            match c.keyword() {
+                Some(pair) => Ok(pair),
+                None => Err(c.error("expected a keyword")),
+            }
+        })?;
+
+        match keyword {
+            "i32.const" => Ok(AssertExpression::I32(parser.parse::<i32>()?)),
+            "i64.const" => Ok(AssertExpression::I64(parser.parse::<i64>()?)),
+            "f32.const" => Ok(AssertExpression::F32(parser.parse::<NanPattern<Float32>>()?)),
+            "f64.const" => Ok(AssertExpression::F64(parser.parse::<NanPattern<Float64>>()?)),
+            "v128.const" => Ok(AssertExpression::V128(parser.parse::<V128Pattern>()?)),
+            _ => Err(parser.error("expected a [type].const expression"))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+#[allow(missing_docs)]
+pub enum NanPattern<T> {
+    CanonicalNan,
+    ArithmeticNan,
+    Value(T)
+}
+
+impl <'a, T> Parse<'a> for NanPattern<T> where T: Parse<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        if parser.peek::<kw::nan_canonical>() {
+            parser.parse::<kw::nan_canonical>()?;
+            Ok(NanPattern::CanonicalNan)
+        } else if parser.peek::<kw::nan_arithmetic>() {
+            parser.parse::<kw::nan_arithmetic>()?;
+            Ok(NanPattern::ArithmeticNan)
+        } else {
+            let val = parser.parse()?;
+            Ok(NanPattern::Value(val))
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub enum V128Pattern {
+    I8x16([i8; 16]),
+    I16x8([i16; 8]),
+    I32x4([i32; 4]),
+    I64x2([i64; 2]),
+    F32x4([NanPattern<Float32>; 4]),
+    F64x2([NanPattern<Float64>; 2]),
+}
+
+macro_rules! repeat2 {
+    ( $e:expr ) => { $e, $e };
+}
+macro_rules! repeat4 {
+    ( $e:expr ) => { $e, $e, $e, $e };
+}
+macro_rules! repeat8 {
+    ( $e:expr ) => { $e, $e, $e, $e, $e, $e, $e, $e };
+}
+macro_rules! repeat16 {
+    ( $e:expr ) => {{ $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e, $e }};
```

Comment:
The double `{{` may be causing issues here perhaps? Other than that I'm ok with the duplication below, I think it's fine.